[DEPRECATION] Remove ember-metal/map

Serabe opened the pull request:
`ember-metal/map` contains three private classes `OrderedSet`, `Map`,
`MapWithDefault`. They were not used anywhere in the codebase except for tests
or reexports, so they are being deprecated.

Ember Data team has been removing their usages.

Added emberjs/website#2884.

Fixes #15041
Fixes https://github.com/emberjs/ember.js/issues/13815

rwjblue reviewed (commented):
Seems like there is still some usage of this in the wild...

* https://emberobserver.com/code-search?codeQuery=Ember.OrderedSet
* https://emberobserver.com/code-search?codeQuery=Ember.Map
* https://emberobserver.com/code-search?codeQuery=Ember.MapWithDefault

Seems like we should confirm that these usages can be migrated to something else (the ember-data usage seems to be the highest priority), then add a deprecation for using these with an `until: 2.16`...

Serabe commented:
I'll change this to deprecations this weekend :)

Serabe committed:
[DEPRECATION beta] Deprecate Map, MapWithDefault and OrderedSet

This PR deprecates three private classes: Map, MapWithDefault and OrderedSet.
These classes are private and not being used anyplace in this repo's code.
Though they are being used in Ember Data, the team is working on removing their
usages.

Serabe commented:
Updated PR. Not sure I approached the deprecation the best way and I for sure have no idea on which the exact URL would be.

Feedback more than welcomed!

homu commented:
:umbrella: The latest upstream changes (presumably #15134) made this pull request unmergeable. Please resolve the merge conflicts.

stefanpenner commented:
> Ember Data team has been removing their usages.

This isn't true, although some scenarios in ED incorrectly used a set, many legitimately require it.

> so they are being deprecated.

I'm alright with deprecating, but there needs to be a transition plan in place before that. As ember-data does require an efficient OrderedSet. Maybe via add-on, or as part of ember-data itself.

Serabe commented:
Rebased.

@stefanpenner we talked about this in #dev-ember and seemed so. In any case, seems more natural to provide an `OrderedSet` from an add-on or via EmberData itself, given it is not being used in any place in this repo. Happy to move the code and tests to a new repo as part of this PR.

Serabe commented:
Closing this.
